### DISCUSSION TITLE
Custom Array Type filtering is broken

### DISCUSSION THREAD

**Discussion by maphe** [mikro-orm/mikro-orm#1212] [Q&A] [answered]
**Describe the bug**
I created a custom type and when trying to filter on that field the ORM transforms my filter in a weird way.

**To Reproduce**
I created a spec here: https://gist.github.com/maphe/a2b02158e1dbfcb0048d890c8b4a6615

**Expected behavior**
Basically my custom type mimics a Ruby Rails array since we have to deal with some legacy DB, so a typical array
```
['a', 'b', 'c']
```
represents as a blob
```
---
- a
- b
- c
```
The custom type works fine for inserts and selects, but filters don't. I suspect this has to do with the property type hint being an array.

I kinda want to simulate the `$in` operator like `'engineering' in teams`. This find condition: `{ teams: { $like: '%- engineering%' } }`
turns into the following SQL:
```sql
select `e0`.* from `legacy_user` as `e0` where `e0`.`teams` like '---- %- -- - e- n- g- i- r'
```

Please let me know if I'm doing something wrong or if it's a bug on the ORM that can get fixed. Thanks

**Versions**

| Dependency | Version |
| - | - |
| node | 12 |
| typescript | 4.1.2 |
| mikro-orm | 4.3.4 |
| your-driver | sqlite/mysql |

**Comment by B4nan** [comment]
So what query you actually want? 

The behaviour seems correct, custom type value is always converted to the db type, even when filtering. You can't just use it as string, as you told to the ORM that it is not string, but rather a `RailsArrayType`.

You can try disabling processing of customm types via `FindOptions`:

```ts
const users = await orm.em.getRepository(LegacyUser).find({ teams: { $like: '%- engineering%' } }, { convertCustomTypes: false });
```

**Comment by maphe** [comment]
@B4nan thanks for the tip, couple of questions though:
1. how to apply the same option to a query builder?
2. if I had 2 different custom types on 2 columns in the same query, there would be no way to let one auto-convert and turn it off for the other one?

**Comment by B4nan** [comment, marked as answer]
1. `qb.unsetFlag(QueryFlag.CONVERT_CUSTOM_TYPES)`
2. nope, you cant have mixed query, either all db values or all runtime values

**Comment by maphe** [comment]
I might be something wrong but it's not working:
```ts
      const users   = await orm.em.getRepository(LegacyUser)
        .find(
          { teams: { $like: '%- engineering%' } },
          { convertCustomTypes: false },
        );
      const usersQb = await orm.em.getRepository(LegacyUser).createQueryBuilder('u')
        .select('*')
        .where({ teams: { $like: '%- engineering%' } })
        .unsetFlag(QueryFlag.CONVERT_CUSTOM_TYPES)
        .getResult();
```

print out following SQLs
```sql
[query] select `e0`.* from `legacy_user` as `e0` where `e0`.`teams` like '%- engineering%' [took 0 ms]
[query] select `u`.* from `legacy_user` as `u` where `u`.`teams` like '---- %- -- - e- n- g- i- r' [took 0 ms]
```

**Reply by B4nan** [reply]
try unsetting the flag before you set the query

```ts
      const usersQb = await orm.em.getRepository(LegacyUser).createQueryBuilder('u')
        .unsetFlag(QueryFlag.CONVERT_CUSTOM_TYPES)
        .select('*')
        .where({ teams: { $like: '%- engineering%' } })
        .getResult();
```

**Reply by maphe** [reply]
Apparently doing the following works as well:
```ts
      const usersQb = await orm.em.getRepository(LegacyUser).createQueryBuilder('u')
        .select('*')
        .unsetFlag(QueryFlag.CONVERT_CUSTOM_TYPES)
        .where({ teams: { $like: '%- engineering%' } })
        .setFlag(QueryFlag.CONVERT_CUSTOM_TYPES)
        .getResult();
```
so I'm guessing that would be a way to mix converted and not converted custom types in a same query (turning the flag on and off per where clause). Although still not achievable via `FindOptions`.

**Comment by maphe** [comment]
Ah thanks, working now.

Not sure how doable that would be but I'd suggest a method in the `Type` abstract class that would handle the conversion if any conversion is needed. Something like:
```ts
buildFilter(operator, values: JSType): QBFilterQuery
```
and so the default would just call `convertToJSValue` but we could override, for instance in my case
```ts
buildFilter(operator, values: JSType): QBFilterQuery {
  switch(operator) {
    case '$eq':
      return { $like: '%- engineering%' };
      break;
    default:
      return this.convertToJSValue(values);
  }
}
```
